src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/DecryptorPalWindows.DecodeRecipients.cs

```diff
@@ -0,0 +1,77 @@
+    {
+        private static RecipientInfoCollection CreateRecipientInfos(SafeCryptMsgHandle hCryptMsg)
+        {
+            int numRecipients;
+            int cbRecipientsCount = sizeof(int);
+            if (!Interop.Crypt32.CryptMsgGetParam(hCryptMsg, CryptMsgParamType.CMSG_CMS_RECIPIENT_COUNT_PARAM, 0, out numRecipients, ref cbRecipientsCount))
+                throw Marshal.GetLastWin32Error().ToCryptographicException();
+
+            List<RecipientInfo> recipientInfos = new List<RecipientInfo>(numRecipients);
+            for (int index = 0; index < numRecipients; index++)
+            {
+                // Do not dispose this safehandle. The RecipientInfoPal(s) objects we create hold on to these and own the job of disposing them.
+                SafeHandle pCmsgCmsRecipientInfoMemory = hCryptMsg.GetMsgParamAsMemory(CryptMsgParamType.CMSG_CMS_RECIPIENT_INFO_PARAM, index);
+                IEnumerable<RecipientInfo> recipientInfosForThisIndex = ToRecipientInfosForThisIndex(pCmsgCmsRecipientInfoMemory, index);
+                recipientInfos.AddRange(recipientInfosForThisIndex);
+            }
+
+            return new RecipientInfoCollection(recipientInfos);
+        }
+
+        private static IEnumerable<RecipientInfo> ToRecipientInfosForThisIndex(SafeHandle pCmsgCmsRecipientInfoMemory, int index)
+        {
+            bool mustRelease = false;
+            pCmsgCmsRecipientInfoMemory.DangerousAddRef(ref mustRelease);
+            try
+            {
+                unsafe
+                {
+                    CMSG_CMS_RECIPIENT_INFO* pCMsgCmsRecipientInfo = (CMSG_CMS_RECIPIENT_INFO*)(pCmsgCmsRecipientInfoMemory.DangerousGetHandle());
+                    switch (pCMsgCmsRecipientInfo->dwRecipientChoice)
+                    {
+                        case CMsgCmsRecipientChoice.CMSG_KEY_TRANS_RECIPIENT:
+                            return new KeyTransRecipientInfo[] { new KeyTransRecipientInfo(new KeyTransRecipientInfoPalWindows(pCmsgCmsRecipientInfoMemory, index)) };
+
+                        case CMsgCmsRecipientChoice.CMSG_KEY_AGREE_RECIPIENT:
+                            {
+                                CMSG_KEY_AGREE_RECIPIENT_INFO* pCmsKeyAgreeRecipientInfo = pCMsgCmsRecipientInfo->KeyAgree;
+                                int numKeys = pCmsKeyAgreeRecipientInfo->cRecipientEncryptedKeys;
+                                KeyAgreeRecipientInfo[] recipients = new KeyAgreeRecipientInfo[numKeys];
+                                for (int subIndex = 0; subIndex < numKeys; subIndex++)
+                                {
+                                    recipients[subIndex] = new KeyAgreeRecipientInfo(new KeyAgreeRecipientInfoPalWindows(pCmsgCmsRecipientInfoMemory, index, subIndex));
+                                }
+                                return recipients;
+                            }
+
+                        default:
+                            throw ErrorCode.E_NOTIMPL.ToCryptographicException();
```

Comment:
We’re matching the desktop behavior in throwing E_NOTIMPL. I think the rationale is that it’s not an “app error” if he tries to decode a ECMS that came across the wire that has an unknown choice value in that field. It’s just as likely that this is a valid new choice that was unknown at the time we implemented EnvelopedCms so it’s “not implemented” (though “not supported” would have been better.)